### ISSUE TITLE
Damoose: The Damus Safari Extension

### ISSUE THREAD

**Issue by jb55** [damus-io/damus#2238]
We can protect our users from bad clients by creating a nostore-like signing extension for browser clients.

We can protect users against buggy clients that nuke contact lists like on:

- https://github.com/nostrability/nostrability/issues/33

### Tracking Issues

- [x] https://github.com/damus-io/damus/issues/2369
- [x] https://github.com/damus-io/damus/issues/2375

**Comment by alltheseas** [comment]
I support this. The issue with browser extensions is that:

1) over 90% of damus users are on iOS,
2) extensions (afaik) are not user friendly on mobile),


Recent learnings for me: the practical experience of trying to set up an extension via a mobile website sucks on iOS, and Android
1) https://github.com/pablof7z/wiki/issues/27
2) https://github.com/pablof7z/wiki/issues/22

However, I know that on Android there is a stand-alone signing extension for mobile called [amber](https://github.com/greenart7c3/Amber).

Could this approach be applied directly to Damus iOS, and later to Damus Android?

**Comment by alltheseas** [comment]
🤔 delegation?

https://github.com/damus-io/damus/issues/451

**Comment by alltheseas** [comment]
Multiple keys on extension would be useful for certain use cases

https://damus.io/nevent1qqsq6enqcl80h0md3s2mw67dp3jm8d4dvj6g7pds5se0gc95gwqqjnqpzemhxue69uhkzarvv9ejumn0wd68ytnvv9hxgqgcwaehxw309aex2mrp0yh8xmn0wf6zuum0vd5kzmqpz4mhxue69uhkummnw3ezu6twdaehgcfwvd3sz9nhwden5te0wfjkccte9ec8y6tdv9kzumn9ws3pc3z2

**Comment by alltheseas** [comment]
Case for damoose to help onboarding to other nostr apps

> One thing I know you said you were looking at - which I still think fits into the onboarding category is making Damus a NIP-46 remote signer.

> Helping new Damus users onboard to not just Damus but other clients too showcases the benefits of nostr more broadly and I think having your main iOS client be your remote signer makes a lot of sense.

https://damus.io/nevent1qqsvnvuayxrxf487amq7qw4gjz4qteau3vks4jq0dj8fx5uwpzmlmwcxrgcv2

**Comment by alltheseas** [comment]
@danieldaquino likes remote signing - web apps, native apps. Signer on phone, or on a server. 

remote signer > extension.

can create own policies for signing (e.g. #2017 )

**Comment by alltheseas** [comment]
The Relationship

  ```
┌─────────────────────────────────────────────────────────────┐
  │                    KEY STORAGE LAYER                        │
  │                                                             │
  │   PR #3470: AccountsStore + KeychainStorage                │
  │   - Stores private keys in iOS Keychain                     │
  │   - Supports iCloud sync OR Secure Enclave encryption       │
  │   - Currently: Only accessible by main Damus app            │
  └─────────────────────────────────────────────────────────────┘
                                │
                                │ App Groups Extension
                                │ (damoose-signer-1x0)
                                ▼
  ┌─────────────────────────────────────────────────────────────┐
  │                   SHARED KEY ACCESS                         │
  │                                                             │
  │   Keys accessible by: Main App + Extensions + Services      │
  └─────────────────────────────────────────────────────────────┘
                                │
                ┌───────────────┴───────────────┐
                ▼                               ▼
  ┌──────────────────────────┐    ┌──────────────────────────┐
  │   SAFARI EXTENSION       │    │   NIP-46 REMOTE SIGNER   │
  │   (NIP-07)               │    │                          │
  │                          │    │                          │
  │ • Injects window.nostr   │    │ • Listens on relays      │
  │ • Signs for WEB clients  │    │ • Signs for ANY client   │
  │ • Local browser only     │    │ • Works remotely         │
  │                          │    │                          │
  │ Examples:                │    │ Examples:                │
  │ - Highlighter            │    │ - Amethyst on Android    │
  │ - Nostrudel              │    │ - Coracle on desktop     │
  │ - Habla.news             │    │ - Any NIP-46 client      │
  └──────────────────────────┘    └──────────────────────────┘
```
  Why [PR #3470](https://github.com/damus-io/damus/pull/3470) is the Foundation

  [PR #3470 ](https://github.com/damus-io/damus/pull/3470)creates the "vault" where private keys are stored:

  // From [PR #3470](https://github.com/damus-io/damus/pull/3470) - AccountsStore.swift
  private func savePrivateKey(_ privkey: Privkey, for pubkey: Pubkey) -> Bool {
      // Stores in Keychain with either:
      // - iCloud sync (kSecAttrSynchronizable: true)
      // - Secure Enclave encryption (local only)
  }

  Problem: This vault is only accessible by the main Damus app. Safari Extensions and background services run in separate sandboxes on iOS.

  What App Groups Enables

  iOS App Groups allow sharing Keychain items between app components:

  // Current (PR #3470) - Main app only
  let query = [
      kSecAttrService: "damus",  // Only main app can access
      ...
  ]

  // With App Groups (damoose-signer-1x0) - Shared access
  let query = [
      kSecAttrAccessGroup: "group.io.damus.shared",  // Extension can access too
      ...
  ]

  How Each Signing Method Works

  Safari Extension (NIP-07) - For Web Clients
```
  User visits highlighter.com
           │
           ▼
  ┌─────────────────────────────┐
  │  highlighter.com JavaScript │
  │                             │
  │  await window.nostr         │
  │    .signEvent(event)        │
  └─────────────────────────────┘
           │
           ▼
  ┌─────────────────────────────┐
  │  Damoose Safari Extension   │
  │                             │
  │  1. Receives sign request   │
  │  2. Shows approval UI       │
  │  3. Reads key from Keychain │◄── Needs App Groups
  │  4. Signs event             │
  │  5. Returns to webpage      │
  └─────────────────────────────┘

  NIP-46 Remote Signer - For Any Client

  User wants to use Amethyst (Android) with Damus keys
           │
           ▼
  ┌─────────────────────────────┐
  │  Amethyst sends request     │
  │  via nostr relay            │
  │                             │
  │  kind: 24133                │
  │  method: "sign_event"       │
  │  encrypted with NIP-44      │
  └─────────────────────────────┘
           │
           ▼ (over relay)
           │
  ┌─────────────────────────────┐
  │  Damoose NIP-46 Service     │
  │  (runs on iPhone)           │
  │                             │
  │  1. Receives relay message  │
  │  2. Decrypts request        │
  │  3. Shows push notification │
  │  4. User approves in app    │
  │  5. Reads key from Keychain │◄── Needs App Groups
  │  6. Signs & sends response  │
  └─────────────────────────────┘
```
  Key Differences

  | Aspect     | Safari Extension (NIP-07) | NIP-46 Remote Signer            |
  |------------|---------------------------|---------------------------------|
  | Clients    | Web browsers only         | Any app, any platform           |
  | Network    | Local (same device)       | Remote (via relays)             |
  | Protocol   | JavaScript injection      | Nostr events (kind 24133)       |
  | Use case   | Browsing nostr web apps   | Cross-device, onboarding others |
  | Connection | Automatic (page load)     | Explicit (bunker:// URL)        |

  Why Both Matter

  From the GitHub issue comments:

  "over 90% of damus users are on iOS... extensions are not user friendly on mobile"

  Safari Extension: Solves the immediate problem of signing for web clients on Safari

  NIP-46: The more powerful solution because:
  - Works for native apps too (not just web)
  - Helps onboard users to other nostr apps
  - "Making Damus a NIP-46 remote signer... showcases the benefits of nostr more broadly"

  TL;DR

  1.[ PR #3470 = Key storage (the vault)](https://github.com/damus-io/damus/pull/3470)
  2. App Groups = Shared access to the vault
  3. Safari Extension = Signs for web clients locally
  4. NIP-46 = Signs for any client remotely

  All three need access to the same keys, which is why PR #3470's infrastructure is the foundation for Damoose.

**Comment by alltheseas** [comment]
On URL scheme handler vs NIP-46 for apps on the same iPhone (e.g. Damus acting as signing app of Primal, Nostur, Comingle etc.)

For multiple iOS nostr apps on the same device, here's the real tradeoff:

  The Core Tradeoff

  |            | URL Schemes         | NIP-46                        |
  |------------|---------------------|-------------------------------|
  | UX Flow    | App switch visible  | Theoretically seamless        |
  | Network    | None needed         | Requires relay connection     |
  | Latency    | ~200ms (app switch) | ~300-800ms (relay round-trip) |
  | Complexity | Simple              | Complex                       |

  But here's the catch on iOS...

  ---
  iOS Background Execution Reality

  NIP-46's main advantage is supposed to be seamless background signing - the signer app handles requests without the user seeing an app switch.

  On Android (Amber): This works great. Amber runs as a background service, receives requests, signs, returns - user never leaves their app.

  On iOS: Apple severely restricts background execution:

  ```
┌─────────────────────────────────────────────────────────────┐
  │                    iOS BACKGROUND LIMITS                    │
  │                                                             │
  │  • Apps suspended ~10 seconds after leaving foreground      │
  │  • No persistent background services (unlike Android)       │
  │  • Background fetch: unreliable, iOS decides when           │
  │  • Push notifications: can wake app, but limited execution  │
  │  • WebSocket connections: killed when app suspended         │
  └─────────────────────────────────────────────────────────────┘
```
  This means NIP-46 on iOS can't be truly seamless for local apps:
```
  NIP-46 on iOS (realistic flow):
  ┌──────────────┐     ┌─────────────┐     ┌──────────────┐
  │ Primal wants │     │   Relay     │     │    Damus     │
  │ to sign      │     │             │     │  (suspended) │
  └──────┬───────┘     └──────┬──────┘     └──────┬───────┘
         │                    │                   │
         │ 1. Send request    │                   │
         │───────────────────▶│                   │
         │                    │ 2. Push notif     │
         │                    │──────────────────▶│ (wakes app)
         │                    │                   │
         │                    │     3. User must tap notification
         │                    │        or open Damus manually
         │                    │                   │
         │                    │◀──────────────────│ 4. Sign & respond
         │◀───────────────────│                   │
         │ 5. Receive signed  │                   │
  └──────┴───────────────────┴───────────────────┴──────────────
```
  Result: User still has to interact with Damus (tap notification) - not much better than URL scheme app switch!

  ---
  Realistic Comparison for iOS

  | Aspect         | URL Schemes                  | NIP-46 (iOS)                      |
  |----------------|------------------------------|-----------------------------------|
  | What user sees | Brief Damus flash, then back | Push notification, tap to approve |
  | Steps for user | 1 (tap approve)              | 2 (tap notif → tap approve)       |
  | Works offline  | ✅ Yes                       | ❌ No                             |
  | Latency        | ~200ms                       | ~500ms + notification delay       |
  | Battery        | Minimal                      | WebSocket + relay overhead        |
  | Reliability    | 100% (direct)                | Depends on relay uptime           |

  ---
  When Each Makes Sense

  URL Schemes: Better for Local iOS Apps

  ✅ Other iOS nostr app on same device
  ✅ Offline usage (airplane mode, poor connection)
  ✅ Speed-critical (rapid posting)
  ✅ Simple "sign this one thing" requests

  NIP-46: Better for Cross-Device/Platform

  ✅ Desktop app wanting iPhone to sign
  ✅ Android app using your iOS keys
  ✅ Web app on laptop using phone as signer
  ✅ Sharing signing ability with trusted device

  ---
  The Pragmatic Answer

  For your specific question (multiple iOS nostr apps):

  URL Schemes wins because:

  1. Same UX outcome - User interaction required either way on iOS
  2. Simpler - No relay dependency, no push notification setup
  3. Faster - Direct app-to-app, no network round-trip
  4. More reliable - No relay downtime concerns
  5. Works offline - Sign posts while on airplane

  NIP-46 is still valuable for:
  - Cross-device signing (the original intent)
  - Web apps (where URL schemes don't work)
  - Future iOS improvements (if Apple relaxes background limits)

  ---
  Recommendation for Damoose

  ```┌─────────────────────────────────────────────────────────────┐
  │  PRIORITY FOR LOCAL iOS APPS                                │
  │                                                             │
  │  1. URL Scheme Handler (nostrsigner://)     ◀── Do first   │
  │     - Fast, reliable, works offline                         │
  │     - Matches Amber's Android approach                      │
  │                                                             │
  │  2. NIP-46 Service                          ◀── Do second  │
  │     - For cross-device/web use cases                        │
  │     - Nice to have, not critical for iOS-to-iOS             │
  └─────────────────────────────────────────────────────────────┘
```

**Comment by alltheseas** [comment]
WIP: https://github.com/damus-io/damus/pull/3506

**Comment by alltheseas** [comment]
Closing as part of a backlog triage — we're narrowing focus to the most actionable issues to reduce noise for contributors.

If this is still relevant, please reopen with updated context or reproduction steps. Thanks for reporting!
